docs: File extension named processor deprecation

matwilko opened the pull request:
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[X] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Added a warning box to the Plugin and Custom Processor docs noting that the file extension-named processor feature has been removed in flat configuration, and therefore will be completely removed in ESLint v9.

Plugin authors need to be aware that when converting to/implementing new flat configs, they can no longer rely on this auto-inclusion, and must explicitly specify their processors in their configs.

nzakas commented:
I'm not sure this wording is correct. Can you link to the reference where you found this info?

In particular, if we haven't made any public announcement about deprecation/removal, then we shouldn't insert it into the docs.

(I'm aware I possibly said something at one point, but I can't find any references on my own, so it would be helpful to link to it for context on this PR.)

matwilko commented:
I hadn't seen any explicit notice about the change in behaviour, which is why I thought a warning needed adding to the docs 😄

It's just something I've run across while setting up a new project using flat config, and I assumed it was a purposeful change as part of moving to flat config - I even had to shim support into my project locally by post-processing the config array and inserting extra elements after elements with plugins that have extension-named processors to support plugins that haven't updated yet.

Given how explicit flat config likes to make things now, only one processor being permitted per-config-element, and nothing else in the system seeming to modify the array by adding/removing elements, it didn't seem odd that this auto-registration hadn't been brought forward, so I didn't think it was a bug.

If this feature wasn't meant to be removed, then this obviously needs to be converted into a bug report 😄

Reproducible example:

```js
// eslint.config.js
var customPlugin = { 
    name: "custom",
    processors: {
        ".js": {
            preprocess: function(text, filename) { console.log('preprocess'); return [{ text: text, filename: "_" + filename }]; },
            postprocess: function(messages) { console.log('postprocess'); return messages[0]; }
        }
    }
};

module.exports = [
    {
        plugins: { customPlugin },
        //processor: customPlugin.processors[".js"]
    }
];
```

Commenting/uncommenting the processor turns it on/off - it's not auto-included in the config. Interestingly enough, you also can't refer to these `.xx` processors by name, `processor: "customPlugin/.js"` results in `TypeError: Key "processor": Expected string in the form "pluginName/objectName" but found "customPlugin/.js"`.

nzakas commented on docs/src/extend/custom-processors.md:
```diff
@@ -139,6 +139,16 @@ See [Specify a Processor](../use/configure/plugins#specify-a-processor) in the P
 ## File Extension-named Processor
 
+::: warning
+With the introduction of [flat configuration](../use/configure/configuration-files-new.md), this feature is being removed. If you are updating your plugin to provide flat configs, you **must** specify all processors explicitly in those configs.
+
+In ESLint v8.0.0+, if a user is still using the old configuration style, processors with file extension names will continue to be automatically included. Users who have switched to using flat configuration may need to manually add these processors to their configuration.
+
+**From ESLint v9.0.0+, when flat configuration becomes required, this feature will be completely removed.**
+
+This feature will continue to be available via `FlatCompat` in [`@eslint/eslintrc`](https://www.npmjs.com/package/@eslint/eslintrc), but there is no intention to continue support for it.
```
I think we can shorten this to remove confusion and ensure people read it. :)

```suggestion
This feature is deprecated and only works in eslintrc-style configuration files. Flat config files do not automatically apply processors; you need to explicitly set the `processor` property.
```

matwilko replied:
Sure, I'll update the other warning in the plugin docs as well to reflect the simplified language.

Is it worth explicitly noting that the feature will be removed in ESLint v9? I was trying to make plugin authors very aware that this change is coming, and if they get bug reports from users switching to v9, they have an easy to spot "oh that's why" reference in the docs.

nzakas commented:
Yes, the file extension processors aren't automatically applied in flat config, but you can manually specify them like `processor: markdown/.md`. I think at some point we were talking about deprecating file extension processors but I can't find a reference, and unfortunately, the team member with the most knowledge about this has left the team.

So, I've left a suggestion for how we can clean up the warning. Please also double-check the linting error.

matwilko commented:
As noted above, you can't refer to extension named processors by name in the config - you have to embed the processor object manually because `"plugin/.xx"` doesn't pass schema validation.

Should I add an extra commit to update the schema validation for `processor`?

matwilko commented:
@nzakas I've updated the language, with an explicit extra warning that the feature will be removed in ESLint v9.

Also added a fix to the processor string validation to allow arbitrary plugin/processor names to be supported.

matwilko commented:
No problem, I'll pull out the code change.

Since you already seem to think the system _should_ allow these processors to be named, but it actually doesn't, I'll post it as a new bug fix PR, and we can discuss there 😄

fasttime commented on docs/src/extend/custom-processors.md:
```diff
@@ -139,6 +139,12 @@ See [Specify a Processor](../use/configure/plugins#specify-a-processor) in the P
+::: warning
+This feature is deprecated and only works in eslintrc-style configuration files. Flat config files do not automatically apply processors; you need to explicitly set the `processor` property.
+
+**From ESLint v9.0.0+, this feature will be removed.**
```
I think that this wording could be confusing because ESLint 9 will still support the eslintrc-style config in legacy mode with `LegacyESLint`. It's the flat config that doesn't support this feature any more.

matwilko replied:
No worries, will do.

I was under the impression that ESLint 9 wouldn't be supporting eslintrc-style _at all_ (except through other tools manually poking at `LegacyESLint` or using `FlatCompat`).

Given that's the impression I've come away with from the docs and announcements (and I like to think I'm a thorough reader!) is there potentially a need to clarify exactly what config system(s) are going to be supported in ESLint 9 and in what ways?

nzakas replied:
The new config system has been in development for years and we've given sporadic updates as we've progressed, and we've reiterated multiple times that v9.0.0 will still support eslintrc. The development plan was laid out in the RFC: https://github.com/eslint/rfcs/tree/main/designs/2019-config-simplification#backwards-compatibility-analysis

Once we have finalized the feature set for v9.0.0, we will also do a larger "What's coming in v9.0.0" blog post where we'll go into more specific details.

fasttime commented:
Maybe it would be good to add a note to the [migration guide](https://github.com/eslint/eslint/blob/main/docs/src/use/configure/migration-guide.md) to explain that extension-named processors won't work any more. The mention that [the syntax for configuring processors hasn't changed](https://github.com/eslint/eslint/blob/main/docs/src/use/configure/migration-guide.md?plain=1#L30) should be updated as well.

matwilko commented:
@nzakas Would you prefer an update to the migration guide in a new PR, or can we roll a further docs update into this PR to keep these related updates together?

nzakas commented:
@matwilko we can definitely add other docs changes related to explaining file-extension processors in this PR. 👍

matwilko commented:
Removed the notice about removal in ESLint 9.

Added a new section to the migration docs covering the differences for processors between eslintrc and flat config, and removed the bullet about `processor` syntax not changing - while the string form is still around, processor objects can also now be embedded directly.

fasttime commented on docs/src/use/configure/migration-guide.md:
```diff
@@ -114,6 +113,99 @@ export default [
+            postprocess() {}
+        },
+        "someProcessor": {
+            proprocess() {},
```
```suggestion
            preprocess() {},
```

fasttime commented on docs/src/use/configure/migration-guide.md:
````diff
@@ -114,6 +113,99 @@ export default [
+As an example with a custom plugin with processors:
+
+```javascript
+// node_modules/eslint-plugin-somePlugin/index.js
````
Since npm package names are not allowed to contain uppercase letters (per https://docs.npmjs.com/cli/v9/configuring-npm/package-json#name), would `eslint-plugin-some-plugin` be a better name?

fasttime commented on docs/src/use/configure/migration-guide.md:
```diff
@@ -114,6 +113,99 @@ export default [
+            postprocess() {}
+        }
+    }
+]
```
```suggestion
];
```

matwilko commented on docs/src/use/configure/migration-guide.md:
````diff
@@ -114,6 +113,99 @@ export default [
+```javascript
+{
+    files: ["**/*.md"]
+    // It's not valid to use the name "somePlugin/.md", so we need to embed the processor directly
````
Yep, it wasn't clear before, so I was going to follow up afterwards, but since there seems to be consensus to allow, I'll remove 😄

fasttime commented on docs/src/use/configure/migration-guide.md:
````diff
@@ -114,6 +113,98 @@ export default [
+
+```javascript
+// eslint.config.js
+import somePlugin from 'eslint-plugin-somePlugin';
````
```suggestion
import somePlugin from "eslint-plugin-someplugin";
```

Using double quotes to keep the style consistent with the rest of the document.

fasttime commented on docs/src/use/configure/migration-guide.md:
```diff
@@ -114,6 +113,98 @@ export default [
+export default [
+    {
+        plugins: { somePlugin },
+        processor: 'somePlugin/someProcessor'
```
```suggestion
        processor: "someplugin/someProcessor"
```

It's better to use double quotes here for consistency.

fasttime commented on docs/src/use/configure/migration-guide.md:
```diff
@@ -114,6 +113,98 @@ export default [
+{
+     overrides: [{
+        files: ["**/*.md"],
+        processor: "somePlugin/.md"
```
```suggestion
        processor: "someplugin/.md"
```

nzakas commented on docs/src/extend/plugins.md:
```diff
@@ -89,6 +89,10 @@ Plugin environments can define the following objects:
 ### Processors in Plugins
 
+::: warning
+File Extension-named Processors are deprecated and only work in eslintrc-style configuration files. Flat config files do not automatically apply processors; you need to explicitly set the `processor` property.
```
Fixing capitalization.

```suggestion
File extension-named processors are deprecated and only work in eslintrc-style configuration files. Flat config files do not automatically apply processors; you need to explicitly set the `processor` property.
```

matwilko replied:
I was matching the capitalization in the section title at [`custom-processors.md#140`](https://github.com/eslint/eslint/blob/53349cae6eb43d0ecb9c267633bf5bcc03660ac1/docs/src/extend/custom-processors.md?plain=1#L140) for consistency, as it seemed like the "proper" name of the feature. Happy to go either way, just pointing out that it was intentional this way 😄

nzakas commented on docs/src/use/configure/migration-guide.md:
```diff
@@ -27,7 +27,6 @@ To use flat config with ESLint v8, place a `eslint.config.js` file in the root o
 While the configuration file format has changed from eslintrc to flat config, the following has stayed the same:
 
 * Syntax for configuring rules
-* Syntax for configuring processors
```
This should not be removed. The syntax for configuring processors is the same (using `processor`).

matwilko replied:
It hasn't stayed the _same_ though, it's expanded. How about if I modified to `Syntax for configuring processors by their name`?

nzakas replied:
That's a detail that isn't important in the migration guide. What we're saying here is that the same way you specified a processor in eslintrc will work in flat config. The goal of this guide is to help people migrate their existing eslintrc configs, so we should tailor all of the information for that purpose. We have a larger config file guide that explores more of the available options.

nzakas commented on docs/src/use/configure/migration-guide.md:
````diff
@@ -114,6 +113,98 @@ export default [
+}
+```
+
+In flat config, you can even directly embed custom processors without an associated plugin:
+
+```javascript
+// eslint.config.js
+
+export default [
+    {
+        processor: {
+            preprocess() {},
+            postprocess() {}
+        }
+    }
+];
+```
````
Let's leave this out. This isn't a tutorial page to show all possible ways to configure flat config, it's a migration guide to go from eslintrc to flat config. Because this isn't a feature that's supported in eslintrc, it shouldn't be mentioned here.

```suggestion
```

matwilko replied:
Sure, it's a new feature in flat config, I wanted to highlight it for users who may have written their own processors locally, and when converting to flat config, can delete the plugin/package that they previously had to wrap the processor in, simplifying that part of their codebase considerably :smile:

fasttime reviewed (approved):
LGTM. Leaving open for @nzakas to verify.

nzakas commented on docs/src/use/configure/migration-guide.md:
```diff
@@ -24,7 +24,7 @@ To use flat config with ESLint v8, place a `eslint.config.js` file in the root o
 ## Things That Haven’t Changed between Configuration File Formats
 
-While the configuration file format has changed from eslintrc to flat config, the following has stayed the same:
+While the configuration file format has changed from eslintrc to flat config, the following will continue to work with no, or only minor, changes:
```
Again, you're confusing the facts here. Let's just keep the existing verbiage.

```suggestion
While the configuration file format has changed from eslintrc to flat config, the following has stayed the same:
```

nzakas commented on docs/src/use/configure/migration-guide.md:
```diff
@@ -114,6 +114,83 @@ export default [
 
+### Processors
+
+In eslintrc files, processors had to be defined in a plugin, and then referenced by name in the configuration. Processors with a name in the special form `.xxx`  were also automatically configured for any files with the extension `xxx`.
```
xxx has some connotations that aren't appropriate for documentation. :smile:

```suggestion
In eslintrc files, processors had to be defined in a plugin, and then referenced by name in the configuration. Processors beginning with a dot indicated a filename extension (such as `.md`) were also automatically configured for any files matching that extension.
```

matwilko replied:
🤦 Whoops, didn't even register, good catch 😄

matwilko replied:
Gone a step further and linked out to the custom processors page so that developers can be sure what we're talking about and get more details:

> Processors beginning with a dot indicated a [file extension-named processor](../../extend//custom-processors.md#file-extension-named-processor) which ESLint would automatically configure for that file extension.

nzakas reviewed (changes requested):
Getting close, just a couple of tweaks to clean things up.

nzakas reviewed (approved):
LGTM. Thanks!
